pilot/pkg/serviceregistry/kube/cache.go

```diff
@@ -62,10 +65,13 @@ func newPodCache(ch cacheHandler) *PodCache {
 
 		ip := pod.Status.PodIP
 
-		log.Printf("Handling event %s for pod %s in namespace %s -> %v", ev, pod.Name, pod.Namespace, ip)
+		log.Debugf("Handling event %s for pod %s in namespace %s -> %v", ev, pod.Name, pod.Namespace, ip)
 
 		if len(ip) > 0 {
 			key := KeyFunc(pod.Name, pod.Namespace)
+			if c != nil {
+				c.EDSUpdater.WorkloadUpdate(ip, pod.ObjectMeta.Labels, pod.ObjectMeta.Annotations)
+			}
```

Comment:
This is very very hacky and going to result in extremely brittle code. Not only that, but it makes it very hard to follow. You might as well add a simple field to the AppendInstanceHandler that denotes whether the instance already exists and is being updated, vs created/deleted.

Reply:
Ack.

Can we decouple the various improvements and changes to interfaces in separate PRs ? It's already getting very big.

Reply:
Sure..